test/dialect/AIEVec/precanonicalization-aieml-llvmir.mlir

```diff
@@ -1,23 +1,43 @@
 // RUN: aie-opt %s -canonicalize-vector-for-aievec="aie-target=aie2 target-backend=llvmir" -split-input-file | FileCheck %s
 
+// CHECK-LABEL: @scalar_extsi_to_broadcast_swap(
+// CHECK-SAME: %[[SIN:.*]]: i8
+func.func @scalar_extsi_to_broadcast_swap(%s: i8) -> vector<32xi32> {
+    // CHECK: %[[SPLAT:.*]] = vector.splat %[[SIN]] : vector<32xi8>
+    // CHECK: %[[EXT:.*]] = arith.extsi %[[SPLAT]] : vector<32xi8> to vector<32xi32>
+    %0 = arith.extsi %s : i8 to i32
+    %1 = vector.broadcast %0 : i32 to vector<32xi32>
+    return %1 : vector<32xi32>
+}
+
+// -----
+
 // CHECK-LABEL: @extsi_to_broadcast_swap(
 // CHECK-SAME: %[[VIN:.*]]: vector<8xi8>
 func.func @extsi_to_broadcast_swap(%v: vector<8xi8>) -> vector<4x8xi32> {
-    // CHECK: %[[BCAST:.*]] = vector.broadcast %[[VIN]] : vector<8xi8> to vector<4x8xi8>
-    // CHECK: %[[EXT:.*]] = arith.extsi %[[BCAST]] : vector<4x8xi8> to vector<4x8xi32>
+    // CHECK: %[[ZV:.*]] = arith.constant dense<0> : vector<4x8xi8>
+    // CHECK: %[[I0:.*]] = vector.insert %[[VIN]], %[[ZV]] [0] : vector<8xi8> into vector<4x8xi8>
+    // CHECK: %[[I1:.*]] = vector.insert %[[VIN]], %[[I0]] [1] : vector<8xi8> into vector<4x8xi8>
+    // CHECK: %[[I2:.*]] = vector.insert %[[VIN]], %[[I1]] [2] : vector<8xi8> into vector<4x8xi8>
+    // CHECK: %[[BC:.*]] = vector.insert %[[VIN]], %[[I2]] [3] : vector<8xi8> into vector<4x8xi8>
```

Comment:
One quick question, I don't see the pattern in this PR that can canonicalize a `vector.broadcast` to four `vector.insert`. Is it  a pattern in our existing codebase or from upstream MLIR? Maybe I'm missing something...

Reply:
I'm using [the upstream lowering of `vector.broadcast`](https://github.com/Xilinx/mlir-aie/blob/ca6c1eaf315055d6624e0fcece1118046d64af4a/lib/Dialect/AIEVec/Transforms/VectorToVectorConversions.cpp#L757), it already does what we need it to do. It also lowers scalar broadcast to splat, hence my changes around that. And I introduced another change that converts "reshape" broadcasts (the ones that simply add leading unit dimensions) because, otherwise, those get converted to inserts that then get canonicalized back to broadcast.